implementation/BeforeInstallPromptEvent.js

```diff
@@ -0,0 +1,154 @@
+/*globals DOMException*/
+"use strict";
+{
+  const internalSlots = new WeakMap();
+  const installProcesses = [];
+  const AppBannerPromptOutcome = new Set([
+    "accepted",
+    "dismissed",
+  ]);
+
+  /**
+   * Emulates the UI thread firing the event.
+   * @param  {BeforeInstallPromptEvent} event
+   * @param  {Element} element A DOM element that simulates the pop-up.
+   * @return {Promise<void>} Resolves when user makes a choice.
+   */
+  async function requestInstallPrompt(event, element) {
+    const internal = internalSlots.get(event);
+    internal.didPrompt = true;
+    const promptOutcome = await showInstallPrompt(element);
+    internal.userChoiceResolver(promptOutcome);
+  }
+
+  /**
+   * Implementation of BeforeInstallPromptEvent.
+   *
+   */
+  class BeforeInstallPromptEvent extends Event {
+
+    constructor(typeArg, eventInit) {
+      // WebIDL Guard. Not in spec, as it's all handled by WebIDL.
+      if (arguments.length === 0) {
+        throw new TypeError("Not enough arguments. Expected at least 1.");
+      }
+      const initType = typeof eventInit;
+      if (arguments.length === 2 && initType !== "undefined" && initType !== "object") {
+        throw new TypeError("Value can't be converted to a dictionary.");
+      }
+      super(typeArg, Object.assign({ cancelable: true }, eventInit));
+
+      if (eventInit && typeof eventInit.userChoice !== "undefined" && !AppBannerPromptOutcome.has(String(eventInit.userChoice))) {
+        const msg = `The provided value '${eventInit.userChoice}' is not a valid` +
+          "enum value of type AppBannerPromptOutcome.";
+        throw new TypeError(msg);
+      }
+      // End WebIDL guard.
+      const internal = {
+        didPrompt: false,
+        userChoice: null,
+        userChoiceResolver: null, // Implicit in spec
+      };
+
+      internal.userChoice = new Promise((resolve) => {
+        if (eventInit && "userChoice" in eventInit) {
+          return resolve(eventInit.userChoice);
+        }
+        internal.userChoiceResolver = resolve;
+      });
+      internalSlots.set(this, internal);
+    }
+
+    prompt() {
+      if (this.isTrusted === false) {
```

Comment:
Well, it's as functional as any other user-constructed event that relies on trust. You are correct that it's crippled - but that's again by design: you can still construct these events and fire them, but it's just "prompt()" that doesn't do anything - but that will be exactly the same for the real ones when constructed by developers. In that sense, what is currently implemented is correct per-spec. 

Note that I could, theoretically, take BeforeInstallPromptEvent.js and plug it into Gecko, because we support JS-based IDL bindings. So, in such an implementation I would get a "trusted" event.